spec.md

```diff
@@ -2241,6 +2255,20 @@ message NodePublishVolumeRequest {
   // This field is OPTIONAL and MUST match the volume_context of the
   // volume identified by `volume_id`.
   map<string, string> volume_context = 8;
+
+  // If SP has VOLUME_MOUNT_GROUP node capability and CO provides
+  // this field then SP MUST ensure that volume is mounted with
+  // provided volume_mount_group and all files and directories
+  // within the volume are readable and writable by the provided
+  // volume_mount_group.
+  // If NodeStageVolume was previously called with volume_mount_group
+  // CO MUST ensure that NodePublishVolume uses the same
+  // volume_mount_group for the same volume_id.
+  // The value of volume_mount_group should be group
+  // identifier (as determined by underlying operating system)
+  // which would be associated with workload that uses the volume.
+  // This is an OPTIONAL field.
+  string volume_mount_group = 9;
```

Comment:
so the next thing that someone will ask for is "mount user". and then someone else will want "mount SELinux labels". and then "extended ACLs". and so on.

at the risk of over-engineering a solution here, i'm wondering if it might make sense to have a top-level type, e.g. `VolumeOwnership` to capture the intent:

```protobuf
message VolumeOwnership {
  string primary_group = 1;
}
```

and then the field in this message becomes:
```protobuf
  VolumeOwnership volume_mount_owner = 9;
```

Reply:
The concern about `mount_user` is valid. We may need that in future. SELinux however is solvable without any CSI spec change - see - https://github.com/kubernetes/enhancements/blob/master/keps/sig-storage/1710-selinux-relabeling/README.md

Reply:
Also the reason we need this as explicit field in `nodestage/nodepublish` is because - unlike selinux mount options which has a standard key/value pair (like `mount -o context=<>`), the group mount option has no such standard format. So CO does not know how to format it.

Reply:
Did we conclude on this? I think it would be nice to leave room for future options.

Reply:
hmm, so I was trying to update the spec for moving `volume_mount_group` to its own message type, but there are some issues to think about. 

Currently the spec says, if volume was staged with one group identifier, it can't be node published with different group identifier on same node. This makes sense for group identifier but what if in future it does not make sense for uid or  something else? 

Using SELinux as an exmaple -if we were to add `selinux_context` to `VolumeOwnership` message then, it would be an error to NodePublish a volume with different selinux contexts on same node, it just won't work. But it **does** work for group identifier for many CIFS implementations at least.

So bringing all of ownership related options within one message can also complicate such future changes.

Reply:
Does the restriction have to apply at the higher level VolumeOwnership or can it apply on individual fields?

Reply:
I would expect restrictions to apply to individual fields but error handling and recovery has to be appropriately adjusted so as it can apply to individual fields.

Reply:
The trouble is basically - appropriately wording `VolumeOwnership` type which can apply to both NodeStage and NodePublish and then defining errors and recovery for both stage and publish in same place and type and then ensuring that error handling and recovery appropriately applies to all the fields and is easy to extend in future. 

This IMO leads to excessively long definition of type `volume_mount_group`  (because among other things now we have to define both publish and stage time behaviour) for example. 

It is not impossible but I am not sure if this is worth the tradeoff. @jdef @msau42 what do you think?

Reply:
we could consider other naming:
```proto
message VolumePrincipal {
  string group = 1;
}
```

Reply:
we can certainly do this. Did we want this as part of `VolumeCapability` field?